### PR TITLE
Add page manager

### PR DESCRIPTION
**Pull request by Zefinder**
Adds a page manager to the OS so the task get assigned to it. The kernel and user PGD are also automatically filled

### CODE REVIEW

File: kernel/core/extend/page_manager.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+static struct task_pages_t task_pages[TASK_NUMBER];
+
+// Inits the page manager
+void init_page_manager() {
+    // Inits page address stack
+    init_page_stack();
+
+    // Put task_pages to 0
+    memset(task_pages, 0, sizeof(struct task_pages_t) * TASK_NUMBER);
```
Comment (angeredShrimp) [inline kernel/core/extend/page_manager.c]:
```suggestion
    memset(task_pages, 0, sizeof(task_pages));
```

File: kernel/core/extend/page_manager.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    // If no pages yet, it means that we need to link the PTB to the PGD for the good index
+    if (task_pages[task_index].number_user_pages == 0) {
+        link_ptb(task_index);
+    }
+
+    // Pop address from stack
+    offset_t address = pop_page_address();
+    
+    // If address is -1 then return -1
+    if (address == (offset_t) -1) {
```
Comment (angeredShrimp) [inline kernel/core/extend/page_manager.c]:
```suggestion
    if (address == (offset_t)-1) {

```

Comment (angeredShrimp) [reply kernel/core/extend/page_manager.c]: Actually, return NULL instead of -1 for addresses and offsets

File: kernel/core/extend/page_manager.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        // TODO Unlink PTB
+        return -1;
```
Comment (angeredShrimp) [inline kernel/core/extend/page_manager.c]: Return NULL

File: kernel/core/extend/stacks.c [hunk 4]
```diff
@@ -63,3 +88,10 @@ pte32_t* pop_ptb() {
     // Else we return the ptb and we increase the ptb_stack_index
     return ptb_stack_index == NUM_PTB ? NULL : ptb_stack[ptb_stack_index++];
 }
+
+// Pops the page address that is ready to use, removing it from the stack. If none returns -1
+offset_t pop_page_address() {
+    // If no value is pointed (after PAGE_NUMBER - 1), we return -1
+    // Else we return the page address value and we increase the page_stack_index
+    return page_stack_index == PAGE_NUMBER ? (offset_t) -1 : page_stack[page_stack_index++];
```
Comment (angeredShrimp) [inline kernel/core/extend/stacks.c]: Return NULL instead of -1

File: unit_test/Makefile [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 CC         := gcc
 INCLUDE    := -I../kernel/include
-CFLAGS	   := -pipe -nostdlib -nostdinc -ffreestanding -fms-extensions
+CFLAGS	   := -pipe -nostdlib -nostdinc -ffreestanding -fms-extensions -Wno-int-to-pointer-cast
```
Comment (angeredShrimp) [inline unit_test/Makefile:5]: Remove because we return NULL instead of -1 for addresses and offsets

File: unit_test/test_script.sh [hunk 6]
```diff
@@ -10,7 +10,7 @@ test_file="test"
 filenames=($(ls | grep .c | cut -d' ' -f1))
 
 CC="gcc"
-OPTIONS="-fms-extensions -rdynamic -finstrument-functions"
+OPTIONS="-fms-extensions -rdynamic -finstrument-functions -Wno-builtin-declaration-mismatch -Wno-pointer-to-int-cast -Wno-int-to-pointer-cast"
```
Comment (angeredShrimp) [inline unit_test/test_script.sh]: skreugneugneu

**Review by angeredShrimp — CHANGES_REQUESTED** [review]
Please it would be very much better to return NULL when there is an invalid pointer instead of a cast -1, even if we have to add an offset to virtual addresses

File: unit_test/Makefile (hunk 5 above)
Comment (Zefinder) [reply unit_test/Makefile:5]: In `stack.c`:
```c
for (int i = 0; i < NUM_PTB; i++) {
   ptb_stack[i] = (pte32_t*)(PTB_START + i*PTB_SIZE);
}
```
This causes an Wint-to-pointer-cast, thus not my problem

**Review by angeredShrimp — APPROVED** [review]

**Pull request merged**